File: src/bin/84.largest-rectangle-in-histogram.py
#
# @lc app=leetcode.cn id=84 lang=python3
# @lcpr version=30204
#
# [84] 柱状图中最大的矩形
#


# @lcpr-template-start
from typing import List, Optional
from src.adt.py.leetcodeType import ListNode, TreeNode
# @lcpr-template-end
# @lc code=start
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        '''
        for pillar at index i

        let right be the index of first smaller value
        same for left     

        max area = (right - left) * height[i]
        '''

        n = len(heights)
        
        leftBound = [-1] * n
        st = []
        for i in range(n):
            while st and heights[i] < heights[st[-1]]:
                st.pop()

            if st:
                leftBound[i] = st[-1]
            
            st.append(i)

        rightBound = [n] * n
        st = []
        for i in range(n):
            while st and heights[i] < heights[st[-1]]:
                rightBound[st.pop()] = i
            
            st.append(i)

        area = [0] * n
        for i in range(n):
            area[i] = (rightBound[i] - leftBound[i] - 1) * heights[i]

        return max(area)


# @lc code=end



#
# @lcpr case=start
# [2,1,5,6,2,3]\n
# @lcpr case=end

# @lcpr case=start
# [2,4]\n
# @lcpr case=end

#

File: src/warehouse/84.largest-rectangle-in-histogram.java
/*
 * @lc app=leetcode.cn id=84 lang=java
 * @lcpr version=30204
 *
 * [84] 柱状图中最大的矩形
 */


// @lcpr-template-start

// @lcpr-template-end
// @lc code=start

import java.util.Arrays;
import java.util.Stack;

class Solution {
    public int largestRectangleArea(int[] heights) {
        // area = (r - l) * height[i]

        int[] leftSmall = new int[heights.length];
        int[] rightSmall = new int[heights.length];

        Stack<Integer> st = new Stack<>();
        for (int i = 0; i < heights.length; i++) {
            while (!st.isEmpty() && heights[i] <= heights[st.peek()]) {
                st.pop();
            }
            if (!st.isEmpty()) {
                leftSmall[i] = st.peek();
            } else {
                leftSmall[i] = -1;
            }
            st.add(i);
        }

        st.clear();
        for (int i = heights.length - 1; i >= 0; i--) {
            while (!st.isEmpty() && heights[i] <= heights[st.peek()]) {
                st.pop();
            }
            if (!st.isEmpty()) {
                rightSmall[i] = st.peek();
            } else {
                rightSmall[i] = heights.length;
            }
            st.add(i);
        }

        // Arrays.stream(leftSmall).forEach(System.out::println);
        // System.out.println();
        // Arrays.stream(rightSmall).forEach(System.out::println);

        int res = 0;
        for (int i = 0; i < heights.length; i ++) {
            int area = (rightSmall[i] - leftSmall[i] - 1) * heights[i];
            res = Math.max(res, area);
        }

        return res;
    }
}
// @lc code=end



/*
// @lcpr case=start
// [2,1,5,6,2,3]\n
// @lcpr case=end

// @lcpr case=start
// [2,4]\n
// @lcpr case=end

 */

